### ISSUE TITLE
np.copy() does not  copy the dtype

### ISSUE THREAD

**Issue by timhoffm** [numpy/numpy#12817]
### Reproducing code example:
~~~
>>> import numpy as np
>>> a = np.zeros((3, 3), dtype=[('x', float), ('y', float)])
>>> b = np.copy(a)
>>> a.dtype is b.dtype
True
~~~

This is problematic because I can change `b.dtype.names`, which implicitly also changes `a.dtype.names`:
~~~
>>> b.dtype.names = ('u', 'v')
>>> print a.dtype.names
('u', 'v')
~~~
This is unexpected since I assume that the array returned by `np.copy()` should not be connected to the input array in any way.

**Comment by mattip** [comment]
If they are immutable, we should disallow writing to names

**Comment by seberg** [comment]
You can write to names? Then they are mutable I guess, and we probably should copy it. However, I do think that conceptionally they should be immutable, and we should probably deprecate writing to names?

**Comment by mattip** [comment]
on the other hand, this is like a `copy.copy`, rather than a `copy.deepcopy` Maybe we should have a `ndarray.deepcopy` that would also copy any possible `object`s in array

**Comment by mattip** [comment]
.. as well as the `dtype`

**Comment by eric-wieser** [comment]
+1 from me on deprecating name assignments

**Comment by charris** [comment]
> If they are immutable, we should disallow writing to names

Making them immutable is a goal, but we haven't been able to reach it yet. Discussions about that first started when we were fixing dtype hashes.

**Comment by mhvk** [comment]
Indeed, `deepcopy` is already more or less what `np.copy` is meant to implement (a regular copy would be more like a new view).

Anyway, I also agree we should try to make the dtype properly immutable; using a dtype with different names can be done with `.view`.

**Comment by eric-wieser** [comment]
Is it worth sticking in a deprecation warning early in the 1.17 cycle, and seeing if anyone is actually using it?

**Comment by timhoffm** [comment]
> using a dtype with different names can be done with .view.

So renaming fields would require
~~~
newnames = ('u', 'v')
a = a.view(dtype=[(newname, type_[0])
           for newname, (name, type_) in zip(newnames, a.dtype.fields.items())])
~~~
or is there a simpler solution?

**Comment by eric-wieser** [comment]
@mhvk: That's incorrect. Those aren't offsets, those are subarray shapes.

**Comment by timhoffm** [comment]
@eric-wieser Thanks for the explanation. I see that this is non-trivial, so some sort of helper function would be nice.

Maybe add this as a method to `dtype` such that one can obtain a new dtype using `dtype.renamed(newnames)` or `dtype.renamed(x='u', y='v')`.

**Comment by eric-wieser** [comment]
@mhvk: I think we deprecated that construct, as prior to python 3.6 it would generate the fields in an unspecified order.

**Comment by seberg** [comment]
It would add a method that just errors on some dtypes (unless we make it structured dtype specific), something like `dtype.renamed(old=new)` or similar. Tagging it to the array could be possible, but doesn't sound clean.
It is likely not  good idea, but I am not sure how often such renaming exists... My guess would be that most people really doing such things tend to use pandas, but I don't know.

**Comment by timhoffm** [comment]
> It is likely not good idea, but I am not sure how often such renaming exists... My guess would be that most people really doing such things tend to use pandas, but I don't know.

One use case would be renaming an array created with `np.genfromtxt(names=True)`. There are cases in which you have really numeric-only tables and don't want/need pandas.

**Comment by seberg** [comment]
See also gh-1725 about immutability of dtypes. Should we put this on the immediate agenda? We would have to provide a way forward for those actually assigning to names. Would that `renamed` be a reasonable approach? (at least to get an alpha version to discuss on the list)

**Comment by mhvk** [comment]
Wow, that goes back quite a bit!

I'm slightly uncomfortable with adding a `renamed` method to the `dtype` class, especially as we're trying to redesign it. Might there be a way to pass a base dtype into `np.dtype` for which one can change properties, say,
```
np.dtype({'base': old_dtype, 'names': new_names})
```

**Comment by seberg** [comment]
Frankly, I am not too uncomfortable with additional methods. If we split out dtypes into their own classes/object hierarchy it seems natural to allow dtype specific methods. Say for example `unit[feet].to_si_unit()`?

**Comment by mhvk** [comment]
@seberg - I agree that in the new system we may well want those methods. I mostly wondered if it wasn't better not to change things while we are trying to move to a new system!

p.s. Indeed, in `astropy`, we do have an `.si` property that returns a new unit:
```
In [9]: unit = u.Unit('AU/yr')

In [10]: unit
Out[10]: Unit("AU / yr")

In [11]: unit.si
Out[11]: Unit("4740.47 m / s")
```

**Comment by seberg** [comment]
Ah, I guess I am more worried about the current mutability creating possible issues ;).

**Comment by mhvk** [comment]
We can ensure `names` become immutable now, I think? After all, there are ways to get a dtype with new names, even if inconvenient?

**Comment by seberg** [comment]
@mhvk yeah, maybe that is the way to go, if it only hits one or two people, we can expect them to add their own small helper if necessary. If it is used more, it feels possibly too inconvenient though.
